### PR TITLE
Warn if an item coming from more recent version than MSRV is used 

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of https://github.com/rust-lang/rust-clippy/issues/6324.

~~Currently, the lint is not working for the simple reason that the `stable` attribute is not kept in dependencies. I'll send a PR to rustc to see if they'd be okay with keeping it.~~

EDIT: There was actually a `lookup_stability` function providing this information, so all good now!

cc @epage

changelog: create new [`incompatible_msrv`] lint

### CODE REVIEW

**Comment by blyxyas** [comment]
I'm not sure if this should be on `suspicious` or `correctness`. A MSRV setting is very important for people that take them into account.

Although I'm not sure why Clippy has a MSRV concept though, as if a project has a MSRV CI would already build it with that Rust version... But those are philosophical questions we don't do (・∀・)

**Comment by epage** [comment]
> Although I'm not sure why Clippy has a MSRV concept though, as if a project has a MSRV CI would already build it with that Rust version... But those are philosophical questions we don't do (・∀・)

For myself, I see this being useful for catching problems in development (since enough people run clippy locally, whether directly or via editor plugins) rather than waiting until a PR is posted and waiting for the CI run to fail.

**Comment by GuillaumeGomez** [comment]
Same as @epage for me. I can think of one extra case: people not ensuring that in CI (bad practice but I think it's definitely not uncommon) and who discover the MSRV breakage when they try to publish.

File: clippy_lints/src/incompatible_msrv.rs
Comment (blyxyas) [inline clippy_lints/src/incompatible_msrv.rs:72]: Meow (does this work?)

```suggestion
                } => Some(version),
```

File: clippy_lints/src/incompatible_msrv.rs
Comment (blyxyas) [inline clippy_lints/src/incompatible_msrv.rs:59]: Thanks for this light caching, it does indeed work! yeehaw :cowboy_hat_face:

File: clippy_lints/src/incompatible_msrv.rs
Comment (blyxyas) [inline clippy_lints/src/incompatible_msrv.rs:117]: You can use `extract_msrv_attr!`, if you want :3 although this approach doesn't disgust me.

File: clippy_lints/src/incompatible_msrv.rs
Comment (blyxyas) [inline clippy_lints/src/incompatible_msrv.rs:38]: So... What do you think about putting this lint in `correctness`?

**Review by blyxyas — APPROVED** [review]
Very good first iteration! Meow meow, just a few nits and questions, but I'll approve this one already (it *could* be merged in this state, but I just want to clarify a few things)

File: clippy_lints/src/incompatible_msrv.rs
Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs:59]: :smiley:

File: clippy_lints/src/incompatible_msrv.rs
Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs:72]: Doesn't work unfortunately:

```
error[E0308]: `if` and `else` have incompatible types
  --> clippy_lints/src/incompatible_msrv.rs:72:16
   |
62 |            let version = if let Some(version) = tcx
   |  ________________________-
63 | |              .lookup_stability(def_id)
64 | |              .and_then(|stability| match stability.level {
65 | |                  StabilityLevel::Stable {
...  |
71 | |              version
   | |              ------- expected because of this
72 | |          } else if let Some(parent_def_id) = tcx.opt_parent(def_id) {
   | | ________________^
73 | ||             self.get_def_id_version(tcx, parent_def_id)
74 | ||         } else {
75 | ||             RustcVersion::new(1, 0, 0)
76 | ||         };
   | ||         ^
   | ||_________|
   |  |_________`if` and `else` have incompatible types
   |            expected `RustcVersion`, found `rustc_semver::RustcVersion`
   |
   = note: `rustc_semver::RustcVersion` and `RustcVersion` have similar names, but are actually distinct types
note: `rustc_semver::RustcVersion` is defined in crate `rustc_semver`
  --> /home/imperio/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustc-semver-1.1.0/src/lib.rs:58:1
   |
58 | pub enum RustcVersion {
   | ^^^^^^^^^^^^^^^^^^^^^
note: `RustcVersion` is defined in crate `rustc_session`
```

File: clippy_lints/src/incompatible_msrv.rs
Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs]: Done. :+1:

File: clippy_lints/src/incompatible_msrv.rs
Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs]: Knowing about this would have saved me so much XD

**Comment by GuillaumeGomez** [comment]
Applied suggestions (thanks!) and replied to questions.

File: clippy_lints/src/incompatible_msrv.rs
Comment (epage) [inline clippy_lints/src/incompatible_msrv.rs:67]: In case `stable` gets stabilized, should we guard this to be exclusive to certain crates?  Or should we wait until that wished for day to arrive?

Comment (epage) [reply clippy_lints/src/incompatible_msrv.rs:67]: (and yes, when its stabilized, I would love to tell clippy or rustc the minimum version for every dependency and have a lint to catch crate dependencies and not just std/alloc/core)

File: clippy_lints/src/incompatible_msrv.rs
Comment (epage) [inline clippy_lints/src/incompatible_msrv.rs:102]: Should we spell out MSRV for those less familiar?

Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs:102]: Let's do that. :+1:

File: clippy_lints/src/incompatible_msrv.rs
Comment (epage) [inline clippy_lints/src/incompatible_msrv.rs:36]: I'm a bit hesitant about using "msrv" because (1) it is an acronym and (2) not everyone knows what it means.

Would "incompatible_rust_version` work for a lint name?

Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs:36]: Hum... What do you think about `incompatible_minimum_rust_version`?

Comment (epage) [reply clippy_lints/src/incompatible_msrv.rs:36]: No meaningful preference either way.

Comment (GuillaumeGomez) [reply clippy_lints/src/incompatible_msrv.rs:36]: I'll go for the more explicit one then.

**Comment by GuillaumeGomez** [comment]
Renamed and update the error message to clear up confusion.

**Comment by blyxyas** [comment]
I'm not sure about the new changes.
If a codebase has a MSRV, then either

1. The codebase is made by 1 author, in which case the MSRV has been set by that author
2. The codebase is made by multiple authors, in which case the concept of an MSRV should be communicated in a contributors guide, by having a mentor tell you that, or seeing the `#![clippy::msrv]` attribute.

Hmmm, the case where spelling it would be useful would be if: An author that has not been communicated of the existence of the MSRV in that codebase (No contributing guide / mentorship), is using a function stabilized later than the MSRV, hasn't seen the `#[clippy::msrv]` attribute and doesn't have access to a search engine / the Rust book.

---

If you want to cover that edge case, we can presume that the same person that wrote `#![clippy::msrv]` also wrote `#![deny(incompatible_msrv)]`. So I'm opposed to changing the lint name, the lint message can be changed with no worries.

**Comment by epage** [comment]
True, `clippy::msrv` uses the acronym as well, which is unfortunate.  Isn't `clippy::msrv` defaulted from `package.rust-version` which most people will be setting instead? And for those users, trying to find a lint for `rust-version` is going to be hampered by the acronym.

I said `clippy::msrv` is unfortunate because acronyms impede communication, especially the more specialized they are.  I expect most people coming from other ecosystems won't be able to recognize MSRV and MSRV won't be covered in most introductory material, so this creates a gap for people's on-boarding journey.  Personally, I feel like this reads a lot less professionally with "MSRV" being used.

**Comment by GuillaumeGomez** [comment]
I removed the last commit (which removed the lint). I also agree with @epage but whatever clippy maintainers prefer. :)

**Comment by blyxyas** [comment]
I agree that acronyms can impede communications, but in this case where we have a `clippy::msrv` attribute / `msrv` configuration and this lint in question is `suspicious` (warn by default).

So the only time where an user would use an attribute would be to either deny the lint or allow it. Both scenarios happen where the user already knows of the lint's existence (and thus, the lint message already explained).

Also, the user can use `cargo clippy --explain <LINT>` at any time, in the case that the person doesn't have access to a search engine.

File: tests/ui/incompatible_msrv.rs
Comment (blyxyas) [inline tests/ui/incompatible_msrv.rs:13]:
```suggestion
    assert_eq!(map.entry("poneyland").key(), &"poneyland");
    //~^ ERROR: is `1.3.0` but this item is stable since `1.10.0`
```

**Review by blyxyas — COMMENTED** [review]
A mini-nit on the test and this is merge-ready!

File: tests/ui/incompatible_msrv.rs
Comment (GuillaumeGomez) [reply tests/ui/incompatible_msrv.rs]: Done!

**Comment by blyxyas** [comment]
=^w^= great! Could you squash these commits?

**Comment by GuillaumeGomez** [comment]
Done as well.

**Review by blyxyas — APPROVED** [review]
LGTM, thanks! :heart:

**Pull request merged**
